decred: Add decred.

JoeGruffins opened the pull request:
part of #986

This pr adds a decred simple payment verification wallet. It does this by taking code at https://github.com/decred/dcrwallet and packaging it with https://github.com/decred/libwallet into a c library that we then use in cake. All private info is kept locally. The wallet will download and verify all block headers and c filters that it uses to find transactions. Thus initial sync can take a bit, depending on bandwidth. After initial sync subsequent syncs and other wallet functions are pretty fast.

There are still some points that need to be resolved that are listed in the body of #986 but this pr is big enough. We hope to get this into master before starting on those.

JoeGruffins commented:
Updated the dcrwallet version and now we can always sync from the wallet birthday which is encoded in the mnemonic.

JoeGruffins commented:
possible bugs found by @ukane-philemon review:

- Importing the same wallet from seed twice:
Importing a wallet whose seed already exists works (for dcr and btc). (intenended?)
The wallet becomes watch-only and users are not able to send from it. However, the send and exchange btns remains active and the user gets an error:

- Cannot view address when losing peers but was synced:
DCR Wallet address is not shown when the wallet is disconnected or it just reconnected but we were already on the view address page. (I think we should still show the wallet address connected or not, btc does)

- Missing tx history:
Missing transaction history: I sent 500 and 45 dcr to the same address at the same time. But got only the transaction history for the first transaction.

- Can't use address book:
Address book is bugged. it creates a new field entry and uses the title as the wallet address and error

JoeGruffins commented:
Another possible huge bug found by @ukane-philemon 

- wallet is not found on ios:
1. I've noticed that creating a dcr wallet on iphone takes a considerable amount of time to finish, ~70sec+.
2. There's something wrong with the wallet data path(as observed on iphone). You create a wallet, restart the app and the wallet dir no longer exists. I've investigated the issue but I've got only some information.
a. Even when a new wallet is created, args["dataDir"] does not exist.
b. args["dataDir"] is the same as walletInfo.dirPath.
```
flutter: create walletflutter: create wallet
flutter: /var/mobile/Containers/Data/Application/4B56953B-9A65-49F4-82B0-6BABCF5D4253/Documents/wallets/decred/Cylindrical Layer # print(args["dataDir"]!);
flutter: false # print(File(args["dataDir"]!).existsSync());
flutter: create wallet end
flutter: /var/mobile/Containers/Data/Application/4B56953B-9A65-49F4-82B0-6BABCF5D4253/Documents/wallets/decred/Cylindrical Layer # print(args["dataDir"]!);
flutter: false # print(File(args["dataDir"]!).existsSync());
flutter: create wallet end
```
After a restart:
```
flutter: load wallet
flutter: /var/mobile/Containers/Data/Application/4B56953B-9A65-49F4-82B0-6BABCF5D4253/Documents/wallets/decred/Cylindrical Layer # print(args["dataDir"]!)
flutter: false # print(File(args["dataDir"]!).existsSync();
flutter: load wallet end
```

exitus1 commented:
@mkyq @OmarHatem28 

Can you guys review this? Decred community is eager to be on Cake.

cyangibm commented:
Hi All, let's get it release please!

gregpent commented:
This would be great for cake wallet and Decred. Can you guys review

jason-at-github commented:
Cake is used by many in the privacy community I can't foresee this integration being anything other than positive for DCR. Looking forward to seeing activity here. Keep up the great work devs and thank you for all you do. We know it isn't easy.

areyoufooled commented:
Decred would be a great addition to Cake wallet

MyWay commented:
Any news on this?

norwnd commented:
Decred community would appreciate a review 🙏

JoeGruffins commented:
Just rebased. Will probably put all suggested changes in a new commit.

JoeGruffins commented on lib/src/widgets/blockchain_height_widget.dart:
```diff
@@ -23,6 +27,7 @@ class BlockchainHeightWidget extends StatefulWidget {
     required this.walletType,
   }) : super(key: key);
 
+  final WalletType? type;
```
Removed the added type variable.

JoeGruffins commented on lib/src/screens/rescan/rescan_page.dart:
```diff
@@ -9,24 +9,51 @@ import 'package:cake_wallet/src/screens/base_page.dart';
+                            RescanWalletState.rescaning,
+                        text: S.of(context).rescan,
+                        onPressed: () async {
+                          await _rescanViewModel.rescanCurrentWallet(
+                              restoreHeight: 0);
```
It will rescan from the wallet birthday which is encoded in the 15 word seed used to create the wallet. This is good as long as no private data is imported into the wallet, which can't currently happen.

JoeGruffins commented on lib/core/wallet_creation_service.dart:
```diff
@@ -109,6 +140,8 @@ class WalletCreationService {
 
   Future<WalletBase> restoreFromSeed(WalletCredentials credentials, {bool? isTestnet}) async {
     checkIfExists(credentials.name);
+    final walletExists = await checkIfWalletWithSeedExists(credentials, credentials.walletInfo!.type);
+    if (walletExists) throw Exception('Wallet with seed already exists!');
```
Removing.

JoeGruffins replied:
Removed.

JoeGruffins commented on lib/core/wallet_creation_service.dart:
```diff
@@ -88,12 +116,15 @@ class WalletCreationService {
 
   Future<WalletBase> restoreFromKeys(WalletCredentials credentials, {bool? isTestnet}) async {
     checkIfExists(credentials.name);
+    final walletExists = await checkIfWalletWithKeyExists(credentials, credentials.walletInfo!.type);
+    if (walletExists) throw Exception('Wallet with public key already exists!');
```
Removed.

JoeGruffins commented on lib/decred/cw_decred.dart:
```diff
@@ -0,0 +1,137 @@
+  }
+
+  @override
+  void updateUnspents(Object wallet) async {}
```
Yeah I guess so... I hadn't notice anything off not doing this but adding.

JoeGruffins commented on lib/decred/cw_decred.dart:
```diff
@@ -0,0 +1,137 @@
+  void updateUnspents(Object wallet) async {}
+
+  @override
+  int heightByDate(DateTime date) {
+    final genesisBlocktime =
+        DateTime.fromMillisecondsSinceEpoch(1454954400 * 1000);
+    final minutesDiff = date.difference(genesisBlocktime).inMinutes;
+    // Decred has five minute blocks on mainnet.
+    // NOTE: This is off by about a day.
+    // TODO: Remove this and just rescan from the wallet birthday.
+    return (minutesDiff / 5).toInt();
```
Yes and no. I don't think we use this anywhere except for rescan, but we are rescanning from birthday now. Removing the TODO and changing the NOTE.

JoeGruffins commented on lib/entities/default_settings_migration.dart:
```diff
@@ -89,6 +90,7 @@ Future<void> defaultSettingsMigration(
               PreferencesKey.currentBalanceDisplayModeKey, BalanceDisplayMode.availableBalance.raw);
           await sharedPreferences.setBool('save_recipient_address', true);
           await resetToDefault(nodes);
+          await setDefaultDecredNodeKey(sharedPreferences, nodes);
```
Ok trying this, let me know if it looks ok.

JoeGruffins commented on lib/view_model/node_list/node_list_view_model.dart:
```diff
@@ -47,6 +49,9 @@ abstract class NodeListViewModelBase with Store {
 
   Future<void> reset() async {
     await resetToDefault(_nodeSource);
+    final decredNode = getDecredDefaultNode(nodes: _nodeSource)!;
+    final sharedPrefs = getIt.get<SharedPreferences>();
+    await setDefaultDecredNodeKey(sharedPrefs, _nodeSource);
```
Thanks, moved down and removed the unneeded code.

JoeGruffins commented on cw_decred/lib/wallet_addresses.dart:
```diff
@@ -0,0 +1,46 @@
+    return addrs;
+  }
+
+  @override
+  set address(String addr) {}
+
+  @override
+  Future<void> init() async {}
+
+  @override
+  Future<void> updateAddressesInBox() async {}
+
+  @override
+  Future<void> saveAddressesInBox() async {}
```
Unsure exactly what's needed here but removing the `saveAddressesInBox` override and calling that in `updateAddressesInBox()` for now.

We should stick unused addresses in `addressesMap`?

JoeGruffins replied:
If I use an address from the address book, it inserts a name. Is this a result of mistakes here?
![image](https://github.com/user-attachments/assets/57e6e1eb-f55d-4c19-aa47-49573fb30fc5)

OmarHatem28 replied:
yup, on this [line](https://github.com/cake-tech/cake_wallet/pull/1322/files#diff-f20234d9bddc253488cfd9721780cb9446ddbb54fc8b7a421bd349e30e9e93feR287)
and if there is any other line that uses the `out.address` it should be like this
```Dart
out.isParsedAddress ? out.extractedAddress! : out.address
```
like this [here](https://github.com/cake-tech/cake_wallet/blob/main/cw_bitcoin/lib/electrum_wallet.dart#L1022)

OmarHatem28 replied:
and for the unimplemented functions, for example, the setter `set address` is going to be used if a user manually selects an address from the address list in the receive screen, you can try it with a bitcoin wallet for example, were you can go to the receive screen, tap on addresses at the bottom, and it will display a list of addresses which are the ones that are used and the ones that are not yet used, and the user can select a different address if he wants

JoeGruffins replied:
Thank you! Will update this area.

JoeGruffins replied:
Is it ok if we don't give out too many unused choices? Because we use an HD wallet, we increase the address index whenever one is used on chain, so if a user were to choose the address at say index 20 when they are at 10, we start to have gaps and "wasted" addresses. We can look at all of these individually, but it takes a bit of back and forth.

The easiest solution is to assume all addresses before the index are used, and all after are unused.

JoeGruffins commented on cw_decred/lib/wallet.dart:
```diff
@@ -0,0 +1,604 @@
+    if (!checkSync()) {
+      return;
+    }
+    final res = libdcrwallet.bestBlock(walletInfo.name);
+    final decoded = json.decode(res);
+    final hash = decoded["hash"] ?? "";
+    if (this.bestHash != hash) {
+      this.bestHash = hash;
+      this.bestHeight = decoded["height"] ?? "";
+    }
```
Oh yeah setting to a string looks like a mistake.

I'm not sure now, I think I planned on using it somewhere in the beginning then didn't. Removing for now.

JoeGruffins commented on cw_decred/lib/wallet.dart:
```diff
@@ -0,0 +1,604 @@
+  @override
+  Future<void> connectToNode({required Node node}) async {
+    if (connecting) {
+      throw "decred already connecting";
```
Ok done.

JoeGruffins commented on cw_decred/lib/wallet.dart:
```diff
@@ -0,0 +1,604 @@
+        syncTimer = null;
+      }
+      persistantPeer = addr;
+      libdcrwallet.closeWallet(walletInfo.name);
+      final network = walletInfo.derivationPath ==
+                  DecredWalletService.seedRestorePathTestnet ||
+              walletInfo.derivationPath ==
+                  DecredWalletService.pubkeyRestorePathTestnet
+          ? "testnet"
+          : "mainnet";
+      libdcrwallet.loadWalletSync({
+        "name": walletInfo.name,
+        "dataDir": walletInfo.dirPath,
+        "network": network,
+      });
+    }
```
The wallet api only allows starting sync, I don't think we can stop it. And we also don't have a command to set peers with the wallet atm. So, need to stop the wallet to stop sync and then start sync again. This could be better but may require changes in dcrwallet. I'll look again to make sure.

JoeGruffins commented on cw_decred/lib/wallet.dart:
```diff
@@ -0,0 +1,604 @@
+  @override
+  Future<void> startSync() async {
+    if (connecting) {
+      throw "decred already connecting";
```
Removing the throw.

JoeGruffins commented on cw_decred/lib/wallet.dart:
```diff
@@ -0,0 +1,604 @@
+    }
+    connecting = true;
+    String addr = "";
+    if (node.uri.host != "") {
+      addr = node.uri.host;
+      if (node.uri.port != "") {
+        addr += ":" + node.uri.port.toString();
+      }
+    }
```
I had problems here when making the address empty when cake saves it. Like, I could not have an empty default node or the app wouldn't start. It was a few months ago so I will look again and elaborate.

By default decred does not need an address or port, we use seeders that are hard coded, so yes empty would be ideal. This is also the first TODO here https://github.com/cake-tech/cake_wallet/issues/986#issuecomment-2109256287

JoeGruffins commented on cw_decred/lib/wallet.dart:
```diff
@@ -0,0 +1,604 @@
+        name: walletInfo.name,
+        peers: persistantPeer,
+      );
+      syncTimer = Timer.periodic(
+          Duration(seconds: 5), (Timer t) => performBackgroundTasks());
```
It is only fetching until it sees one it recognizes, not all. Batches of five.

Five seconds has been good for testing on simnet, but yeah can be whatever you think is best. Setting to 30 seconds.

OmarHatem28 replied:
I get that it's some batches of five, but you still have it in a while true, so it will keep going until it fetches all transactions, which will repeat every 5 seconds, since you set `from` to 0 every time

so why not just save the fetched transactions, and save the last scanned transaction count, and then just try and fetch from that count if you don't find any new transactions it will break out of the loop on the first time

JoeGruffins replied:
I think this is why I originally was saving the block hash so we could skip checking if the block hadn't changed, then I realized that we also want to show pending transactions in the mempool.

Unless I coded something wrong, it should only fetch until it sees a known transaction every time, from the newest, so it would stop at the first batch most of the time. 0 is not a block number if that's what you mean, which is probably where the confusion is from. `this.transactionHistory.update` will return true and the function will return if there is an old tx.

from is the number of transactions to skip, which is not obvious.
```
$ ./dcrctl help listtransactions
listtransactions ("account" count=10 from=0 includewatchonly=false)

Returns a JSON array of objects containing verbose details for wallet transactions.

Arguments:
1. account          (string, optional)                 DEPRECATED -- Unused (must be unset or "*")
2. count            (numeric, optional, default=10)    Maximum number of transactions to create results from
3. from             (numeric, optional, default=0)     Number of transactions to skip before results are created
4. includewatchonly (boolean, optional, default=false) Unused
```

Also this is only a db check, it is not querying other nodes with this call. I do not think checking a number in the db will be faster. Currently this would also mean we have to fetch all transactions every time since I don't think wallet has a method to tell us how many tx there are. I can imagine such a change being met with resistance as well since it doesn't do much.

OmarHatem28 replied:
ok yeah, now it makes sense thanks for elaborating, I thought `from` would start from 0 block height indeed.
we can find a different name for the variable later so it's not confusing

JoeGruffins commented on cw_decred/lib/wallet.dart:
```diff
@@ -0,0 +1,604 @@
+  int walletBirthdayBlockHeight() {
+    final res = libdcrwallet.birthState(walletInfo.name);
+    final decoded = json.decode(res);
+    if (decoded["setfromheight"] == true || decoded["setfromtime"] == true) {
+      return -1;
+    }
+    return decoded["height"] ?? 0;
```
Ah, this is a little confusing, but if those values are set, it means the birthday is not set yet. This would happen during sync, before we find the birthday block.

JoeGruffins replied:
Added a comment.

JoeGruffins commented on cw_decred/lib/wallet.dart:
```diff
@@ -0,0 +1,604 @@
+    var rescanHeight = 0;
+    if (!watchingOnly) {
+      rescanHeight = walletBirthdayBlockHeight();
+      if (rescanHeight == -1) {
+        throw "cannot rescan before the birthday block has been set";
+      }
```
It indicates that sync has not reached the birthday block yet. Can we just return?

JoeGruffins commented on cw_decred/lib/wallet.dart:
```diff
@@ -0,0 +1,604 @@
+      walletId: idPrefix + walletInfo.name,
+      hash: coin.hash,
+      isFrozen: false,
+      isSending: false,
```
I thought isSending meant that the user always wants to use this input in a transaction. Are isSending a pool we should choose from? Then if isSending is false is the same as isFrozen?

JoeGruffins replied:
Doing as you say for now, but I think decred is currently handling this wrong. Right now it is absolutely using all `isSending` and then uses other utxo by selecting the least number that can make the amount we want to send plus fee. It doesn't use `isFrozen` at all.

JoeGruffins replied:
Ok, I think I got it now. `isSending` means we can use but we dont have to. If that is not checked, it is the same as `isFrozen`

OmarHatem28 replied:
And now Decred handles it accordingly?
i.e
for Monero, we pass a list of preferred inputs (which are the ones with `isSending = true`) so that the Monero code uses from this list of inputs while creating the transaction

JoeGruffins replied:
If isSending == false means the same as outputs are frozen, then yes we are doing it correctly. It is only choosing from the `isSending`, i.e., it ignores everything else.

Well, it could be better, since we are telling the wallet which inputs not to use, if I had understood `isSending` better then it would just accept a list of inputs to use. Can we change that later if it seems to be a problem? I can imagine an edge case where a new transaction comes in while creating a transaction, and a blocked is mined at the same time, then we might include that utxo in the transaction although, it would be isSending == true by default, but the user would not have seen it in the list.

JoeGruffins commented on cw_decred/lib/wallet.dart:
```diff
@@ -0,0 +1,604 @@
+            element.hash == coin.hash &&
+            element.vout == coin.vout);
+
+        if (coinInfoList.isEmpty) {
+          this.addCoinInfo(coin);
+        }
```
Ok, copied whats there in xmr.

JoeGruffins commented:
Removed the commit to do with checking if a wallet exists. Addressed most comments as well as I can. Added another commit from wisdom about imports. https://github.com/cake-tech/cake_wallet/compare/5e465c97f635d6dba62f4fcb170a68cfd138f942..05de46bee95532ba5013ecc9831592a31bdab24b

I'm still looking at what we need to do with addresses. We need to supply some number to cake? We can get nearly infinite if we need for something. the internal dcrwallet is handling which are used and won't reuse old ones. Are we showing the old ones somewhere?

OmarHatem28 commented:
> Removed the commit to do with checking if a wallet exists. Addressed most comments as well as I can. Added another commit from wisdom about imports. https://github.com/cake-tech/cake_wallet/compare/5e465c97f635d6dba62f4fcb170a68cfd138f942..05de46bee95532ba5013ecc9831592a31bdab24b
> 
> I'm still looking at what we need to do with addresses. We need to supply some number to cake? We can get nearly infinite if we need for something. the internal dcrwallet is handling which are used and won't reuse old ones. Are we showing the old ones somewhere?
> 
> I hope to add all changes to this last commit so it will probably be updated more.

yup, like mentioned [here](https://github.com/cake-tech/cake_wallet/pull/1322#discussion_r1899252910), we do show old addresses in the receive screen
for the initial generated addresses it depends on how decred would work, but mainly we tend to give the user a few addresses (maybe 10 or 20) so that he can use them willingly before we auto-generate new addresses when needed

JoeGruffins commented on lib/view_model/wallet_address_list/wallet_address_list_view_model.dart:
```diff
@@ -298,9 +314,57 @@ abstract class WalletAddressListViewModelBase
+      return BitcoinURI(amount: amount, address: address.address);
+    }
+
+    if (wallet.type == WalletType.litecoin) {
+      return LitecoinURI(amount: amount, address: address.address);
+    }
+
+    if (wallet.type == WalletType.ethereum) {
+      return EthereumURI(amount: amount, address: address.address);
+    }
+
+    if (wallet.type == WalletType.bitcoinCash) {
+      return BitcoinCashURI(amount: amount, address: address.address);
+    }
+
+    if (wallet.type == WalletType.nano) {
+      return NanoURI(amount: amount, address: address.address);
+    }
+
+    if (wallet.type == WalletType.polygon) {
+      return PolygonURI(amount: amount, address: address.address);
+    }
+
+    if (wallet.type == WalletType.solana) {
+      return SolanaURI(amount: amount, address: address.address);
+    }
+
+    if (wallet.type == WalletType.tron) {
+      return TronURI(amount: amount, address: address.address);
+    }
+
+    if (wallet.type == WalletType.wownero) {
+      return WowneroURI(amount: amount, address: address.address);
+    }
+
+    if (wallet.type == WalletType.decred) {
+      return DecredURI(amount: amount, address: address.address);
+    }
+
+    throw Exception('Unexpected type: ${type.toString()}');
```
Sorry looks like a bad rebase, removed.

OmarHatem28 reviewed (approved):
LGTM so far 🙏 
Thanks for all the effort done on this ^^
will merge it in our repo's decred branch so we can start testing it internally

MrCyjaneK commented:
@JoeGruffins there are `dcrwallet.log` and `external.log`, how can we disable that log files, these are very large in size.
